src/Validation/AbstractValidation.php

```diff
@@ -75,9 +77,70 @@ public function isFileMinSizeValid(int $fileSize, int $minFileSize): bool
 	 */
 	public function isFileTypeValid(string $fileName, string $fileTypes): bool
 	{
-		$fileExtension = explode('.', $fileName);
-		$validTypes = explode(',', str_replace(' ', '', str_replace('.', '', $fileTypes)));
+		$validTypes = $this->parseFiletypesString($fileTypes);
+
+		return in_array($this->getFileExtensionFromFilename($fileName), $validTypes, true);
+	}
+
+	/**
+	 * Checks whether the mimetype for the file is valid,
+	 * i.e. that it matches the extension. If the file is written
+	 * to disk, it'll check its mime_content_type from the filesystem.
+	 * Use the validation filter failMimetypeValidationWhenFileNotOnFS
+	 * to override this behaviour.
+	 *
+	 * @param array<string|int> $file File array.
+	 * @return boolean True if mimetype matches extension, false otherwise.
+	 */
+	public function isMimeTypeValid(array $file): bool
+	{
+		$denyIfFileIsNotUploaded = \apply_filters(Filters::getValidationSettingsFilterName('failMimetypeValidationWhenFileNotOnFS'), false); // phpcs:ignore WordPress.NamingConventions.ValidHookName.NotLowercase
+		$mimeTypes = \array_flip(\wp_get_mime_types());
+
+		$fileMimetype = $file['type'];
+		if ($file['tmp_name'] ?? false) {
+			try {
+				$fileMimetype = \mime_content_type($file['tmp_name']);
+			} catch (\Throwable $t) {
+				if ($denyIfFileIsNotUploaded) {
+					return false;
+				}
+			}
+		} elseif ($denyIfFileIsNotUploaded) {
+			return false;
+		}
 
-		return in_array(end($fileExtension), $validTypes, true);
+		$fileExtension = $this->getFileExtensionFromFilename($file['name']);
+
+		if (\in_array($fileExtension, \explode('|', $mimeTypes[$fileMimetype] ?? []), true)) {
```

Comment:
I'm not a fan of adding extra logic in checks. It makes the code less readable.

I'd `explode` the string first, then check if it's not empty, then do an `isset` check instead of `in_array` 🙂